[inductor] Enable vectorization in fbcode

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105777
* __->__ #105756



In fbcode, to run the test python script (with its accompanying test DSO) we
need to invoke the correct python, with the correct PYTHONPATH, so we supply
those by reading the appropriate values out of `sys`.

It's an improvement for OSS too, since the user may not be running the default
python.

My previous attempt of using `torch.backends.cpu.get_cpu_capability()` didn't work out, for two reasons:
1. That function actually refuses to report AVX512 support; it's #ifdef-ed out, for some reason.
2. In CI, we apparently are picking INVALID_VEC_ISA (at least when running
inductor_timm_cpu_accuracy), whereas `get_cpu_capability` reports AVX2.  This
is surprising, and probably indicates a bug (either in cpu capability or our
test binary), but I'd rather not go digging for it.

Differential Revision: [D47678649](https://our.internmc.facebook.com/intern/diff/D47678649/)

Differential Revision: [D47678649](https://our.internmc.facebook.com/intern/diff/D47678649)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

bertmaher committed:
[inductor] Enable vectorization in fbcode

The method of running a test binary doesn't really work well in fbcode
for various reasons, but there's no reason to do it;
`torch.backends.cpu.get_cpu_capability()` will tell us AVX2 or AVX512 (or
DEFAULT).

I think this might actually be a better method even outside of fbcode, but I
don't have a great way to test AVX512 behavior.

Differential Revision: [D47678649](https://our.internmc.facebook.com/intern/diff/D47678649/)

[ghstack-poisoned]

jansel reviewed (changes requested):
I'd vote for just using `torch.backends.cpu.get_cpu_capability()` always and not diverging behavior based on `is_fbcode()`

bertmaher commented:
Cool.  As long as the run-a-test-script approach doesn’t have some hidden benefit that I’m not seeing, I prefer that approach as well

mikekgfb commented:
A unified approach seems more robust to me, and I think it would be great to have a unified solution.
If we need more build time or hardware time data, we can quite plausibly obtain this by extending the get_cpu_capability with whatever hardware queries are necessary?

bertmaher commented:
> A unified approach seems more robust to me, and I think it would be great to have a unified solution. If we need more build time or hardware time data, we can quite plausibly obtain this by extending the get_cpu_capability with whatever hardware queries are necessary?

I think so.  It seems like a pretty minimalist interface right now, just exposing a few target ISAs.  If we want to get really fancy, it'd be nice to have python bindings for https://github.com/pytorch/cpuinfo (which is the C library we call into to get feature flags)

bertmaher committed:
just use get_cpu_capability on "[inductor] Enable vectorization in fbcode"


The method of running a test binary doesn't really work well in fbcode
for various reasons, but there's no reason to do it;
`torch.backends.cpu.get_cpu_capability()` will tell us AVX2 or AVX512 (or
DEFAULT).

I think this might actually be a better method even outside of fbcode, but I
don't have a great way to test AVX512 behavior.

Differential Revision: [D47678649](https://our.internmc.facebook.com/intern/diff/D47678649/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bertmaher committed:
Update on "[inductor] Enable vectorization in fbcode"


The method of running a test binary doesn't really work well in fbcode
for various reasons, but there's no reason to do it;
`torch.backends.cpu.get_cpu_capability()` will tell us AVX2 or AVX512 (or
DEFAULT).

I think this might actually be a better method even outside of fbcode, but I
don't have a great way to test AVX512 behavior.

Differential Revision: [D47678649](https://our.internmc.facebook.com/intern/diff/D47678649/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bertmaher committed:
ignore this, i just added some logging to understand the crashes on "[inductor] Enable vectorization in fbcode"


The method of running a test binary doesn't really work well in fbcode
for various reasons, but there's no reason to do it;
`torch.backends.cpu.get_cpu_capability()` will tell us AVX2 or AVX512 (or
DEFAULT).

I think this might actually be a better method even outside of fbcode, but I
don't have a great way to test AVX512 behavior.

Differential Revision: [D47678649](https://our.internmc.facebook.com/intern/diff/D47678649/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bertmaher committed:
Update on "[inductor] Enable vectorization in fbcode"


The method of running a test binary doesn't really work well in fbcode
for various reasons, but there's no reason to do it;
`torch.backends.cpu.get_cpu_capability()` will tell us AVX2 or AVX512 (or
DEFAULT).

I think this might actually be a better method even outside of fbcode, but I
don't have a great way to test AVX512 behavior.

Differential Revision: [D47678649](https://our.internmc.facebook.com/intern/diff/D47678649/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bertmaher commented:
Okay, I'm reverting to the subprocess approach; get_cpu_capability is possibly buggy, and I was able to figure out how to get subprocess calls to python working properly @ FB anyways

bertmaher committed:
Update on "[inductor] Enable vectorization in fbcode"




In fbcode, to run the test python script (with its accompanying test DSO) we
need to invoke the correct python, with the correct PYTHONPATH, so we supply
those by reading the appropriate values out of `sys`.

It's an improvement for OSS too, since the user may not be running the default
python.

My previous attempt of using `torch.backends.cpu.get_cpu_capability()` didn't work out, for two reasons:
1. That function actually refuses to report AVX512 support; it's #ifdef-ed out, for some reason.
2. In CI, we apparently are picking INVALID_VEC_ISA (at least when running
inductor_timm_cpu_accuracy), whereas `get_cpu_capability` reports AVX2.  This
is surprising, and probably indicates a bug (either in cpu capability or our
test binary), but I'd rather not go digging for it.

Differential Revision: [D47678649](https://our.internmc.facebook.com/intern/diff/D47678649/)

Differential Revision: [D47678649](https://our.internmc.facebook.com/intern/diff/D47678649)

[ghstack-poisoned]

desertfire commented:
> In CI, we apparently are picking INVALID_VEC_ISA (at least when running inductor_timm_cpu_accuracy)

Hmm, this is a problem. @EikanWang , can you help to investigate?

bertmaher commented:
> > In CI, we apparently are picking INVALID_VEC_ISA (at least when running inductor_timm_cpu_accuracy)
> 
> Hmm, this is a problem. @EikanWang , can you help to investigate?

Here is some example output from a previous version of the diff where I logged what we are picking: https://ossci-raw-job-status.s3.amazonaws.com/log/15275831688

mikekgfb reviewed (approved):
Looks good to me.  Approving.
